src/core/reactor.cc

```diff
@@ -3185,6 +3479,7 @@ int reactor::do_run() {
     std::function<bool()> pure_check_for_work = [this] () {
         return pure_poll_once() || have_more_tasks();
     };
+    _cpu_profiler->start();
```

Comment:
Why is the CPU profiler started and stopped here?

Naively I would expect that it's always running.

Is the period for the CPU profiler to be interpreted in wall clock time or "CPU time"?

Reply:
This is just before the start of the main reactor loop. So I figured starting the profiler here was as good of a place as any. Let me know if you have a better spot in mind though.

The period should be interpreted as "thread time" though with the POSIX timer that won't be a completely accurate statement.

Reply:
Sorry, I was confused, maybe I was referring to the stop()/start() pair around `sleep()` below? Are these calls expensive?

Reply:
I guess there are a few reasons for putting the `start/stop` pair around sleep;
- All of their other interrupt timers follow this pattern.
- We are using `CLOCK_MONOTONIC` now which can cause spurious wake-ups.
- I'm pretty sure, but not entirely convinced that the `PERF_COUNT_SW_TASK_CLOCK`  clock in `perf_event_open` won't cause a spurious wake-up.

Reply:
The cost of the calls depend on the timer chosen. But for the more expensive timer(`perf_event_open`) the number of system calls is as follows;
- `stop`: 1 ioctl
- `start`: 4 ioctl

Reply:
I'm a bit worried about the cost here, especially the latency when waking from sleep, which is an important path (since generally we wake because something occurred like an incoming request). The existing block notifier has an `end_sleep()` call on that path but the method is empty.

What's also worrisome is that sleep() doesn't necessarily sleep: it asks all the pollers whether they can enter "interrupt mode". We miss all this work, which I think is fairly substantial especially for lightly loaded reactors (it's where we do the cross-core assynmetic `membarrier()` for example).

Are "spurious" wakeup a problem? In some way it gives us more information: we sample in "wall clock time units" instead of "reactor running (including polling)" time units, and we will have samples where the reactor is (a) doing using work (b) polling and (c) sleeping. So the tool handling the stack traces can quantify the three types of time, i.e., calculate "reactor utilization". OTOH then I guess the perf-based and timer-based samplers will be inconsistent because you think the SW_TASK_CLOCK doesn't tick when the process is asleep, right?

I wonder if a good approach is a hybrid one: don't disable the timer, but it we wake up during a sleep disable it until the end of the sleep (actually, the signal will end the sleep but we then note this and disable for the next sleep which presumably happens immediately since the wakeup was spurious). This would also catch the work inside `sleep()`.

Reply:
If the profiler is not enabled, the start()/stop() methods are very cheap, right?

Reply:
Both the start/stop check if the profiler is enabled and return if its not. So it's just a branch and some change if it's not enabled.

Reply:
That sounds good but see also the longer previous comment. 

What is your take? Maybe it's not a problem if we are not enabling it by default.

Reply:
Sorry, needed a bit to process the longer comment. Implementing the hybrid approach you suggested wouldn't take long. 

What `SW_TASK_CLOCK` measure exactly lies in the kernel I've found a few descriptions of it's function that are at odds with each other. And I haven't had time to hunt down exactly how it functions in the kernel code yet. For now I'm working under the assumption that it is a measure of thread time and will not increment when said thread is asleep. 

We have metrics for getting the time split between work, polling, and sleeping. Though I can see the benefit in having samples that better explain what each stage is doing. Rather than disabling the profiler after the first sample taken in SS's `sleep()` function it'd be nice to be able to tell the difference between actually sleeping and preparing to sleep. That way we could continue to profile until the thread is actually sleeping.

Reply:
A potential problem with lazily disabling the profiler in `sleep()` is it'd be hard to tell how much of a sample period was actually spent doing work and how much was spent sleeping. We'd probably have to make a call to get the system time before and after the `sleep()` function and compensate in some manner.

Reply:
I.e, say we have a sample period of 100ms and sample for 1s. Then during the sampling we are constantly sleeping for 90ms followed by working for 10ms. We may just get 10 samples from that period of working. The question is then how to interpret those samples.

Reply:
I don't believe introducing a few extra system calls before sleeping when the profiler is enabled is a problem anyhow. If there is so little work to do that the reactor is going to sleep then performance isn't much of a concern. I could be horribly wrong though lol.